### PR TITLE
Adding a tip to run `mix hex.sponsor` when a package that has a sponsor link is added or upgraded

### LINKED ISSUE #883: Show a tip about running mix hex.sponsor whenever a dep with sponsor is added/updated on mix deps.get

### PR DESCRIPTION
**Pull request by kelvinst**
Closes #883

Was not sure where to add tests for it though, but I ran it locally and it printed like this:

Screen Shot 2021-05-14 at 15 37 02

### CODE REVIEW

**Comment by kelvinst** [comment]
Let me know if you want  me to reword it or color the output.

**Comment by kelvinst** [comment]
Actually found out where to put the tests for it, working on that.

**Comment by kelvinst** [comment]
Hey, so I was writing tests for this on `mix_task_test.exs` and was not able to make it pass. To no one's surprise the automated test is actually right, my manual tests were not good enough (I was just `mix deps.unlock`ing a package with a sponsor and `mix deps.get`ing it again, so the dep was downloaded already, only the lock was updated). 

After a while digging it up, found out what is the cause: I added the message to `Hex.RemoteConverger.converge` function, which is called to converge the dep list and the lock to create a new one. Did it because it is already the place that prints the new/upgraded/unchanged dep list, so I thought it makes sense to put it there. 

The problem is that the sponsor link is saved on the package metadata, which is not available before actually downloading the package, so I think the ideal place for the message would be `Hex.RemoteConverger.post_converge`, which would have then to receive the dep list and old lock to recalculate the added/upgraded to show only in the case of added or upgraded sponsored packages.

All that said, the question is: is it worth the trouble? I mean, we would have to change the callback for it on `Mix.RemoteConverger`, maybe it should actually even be a new one to maintain backward compatibility to previous hex versions, IDK. What I'm thinking here is that maybe it would not be a huge hassle if we printed that message everytime one runs `mix deps.get` and has any dep with a sponsor link on the deps list. It would even keep reminding people they rely on packages can be funded, so might even get a better result, IDK 🤷‍♂️

**Comment by kelvinst** [comment]
Committed the failing tests, as I realized they actually work, the problem is on the implementation.

**Comment by ericmj** [comment]
Another option is to add global state via `Hex.State` that records if the message should be printed.

After unpacking [1] we can check for the "sponsor" key and store that in `Hex.State`. `post_converge` will check with `Hex.State` if the message should be printed and reset the value.

It's not pretty but we avoid API changes between Mix and Hex.

@josevalim Any better ideas?

[1] https://github.com/hexpm/hex/blob/main/lib/hex/scm.ex#L165

**Comment by josevalim** [comment]
I am not that familiar with the implementation, so I can't really say which way is best implementation wise. :)

File: lib/hex/remote_converger.ex [hunk 1]
```diff
@@ -236,9 +236,24 @@ defmodule Hex.RemoteConverger do
         unless length(deps) == 0, do: print_category(mod)
         print_dependency_group(deps, mod)
       end)
+
+      if any_sponsored_new_or_updated?(dep_changes) do
+        """
+        You have added or upgraded one or more package that you could sponsor.
+        Run `mix hex.sponsor` to know how.
```
Comment (josevalim) [inline lib/hex/remote_converger.ex:243]:
```suggestion
        You have added/upgraded packages you could sponsor, run `mix hex.sponsor` to know how
```

**Comment by kelvinst** [comment]
I like @ericmj's solution. I mean, as he said, it's not the prettiest, but I don't think this is a feature worth of a change on Mix <> Hex API. Will do that change tomorrow, please let me know if you want me to hold that.

**Comment by kelvinst** [comment]
@ericmj, I have implemented it the way you suggested, works well, do you think we should add more test scenarios?

**Review by ericmj — APPROVED** [review]
This looks great. Thank you @kelvinst!

File: lib/hex/sponsor.ex [hunk 2]
```diff
@@ -0,0 +1,35 @@
+defmodule Hex.Sponsor do
```
Comment (wojtekmach) [inline lib/hex/sponsor.ex:1]:
```suggestion
defmodule Hex.Sponsor do
  @moduledoc false

```

**Review by wojtekmach — APPROVED** [review]

**Comment by wojtekmach** [comment]
Thanks!

**Comment by wojtekmach** [comment]
Oops, I didn't notice there was a problem with the build. I'm gonna revert it, could you re-send the PR?

**Comment by kelvinst** [comment]
Oh, me neither, sure, will resend it this afternoon.

**Comment by kelvinst** [comment]
@wojtekmach, I have reopened it here https://github.com/hexpm/hex/pull/887. Will work on the fixes this week, probably tomorrow.

**Pull request merged**